radio/src/gui/colorlcd/radio_setup.cpp

```diff
@@ -48,151 +48,163 @@ class DateTimeWindow : public FormGroup {
     {
       FormGroup::checkEvents();
 
-      if (get_tmr10ms() - lastRefresh >= 10) {
-        struct gtm newTS;
+      if (seconds && (get_tmr10ms() - lastRefresh >= 10)) {
         lastRefresh = get_tmr10ms();
 
-        gettime(&newTS);
-        if(newTS.tm_sec == lastSecond)
-          return;
-
-        lastSecond = newTS.tm_sec;
-        year->update();
-        month->update();
-        day->update();
-        hour->update();
-        minutes->update();
-        seconds->update();
+        gettime(&m_tm);
+        if (m_tm.tm_year != m_last_tm.tm_year) {
+          m_last_tm.tm_year = m_tm.tm_year;
+          year->update();
+        }
+        if (m_tm.tm_mon != m_last_tm.tm_mon) {
+          m_last_tm.tm_mon = m_tm.tm_mon;
+          month->update();
+        }
+        if (m_tm.tm_mday != m_last_tm.tm_mday) {
+          m_last_tm.tm_mday = m_tm.tm_mday;
+          day->update();
+        }
+        if (m_tm.tm_hour != m_last_tm.tm_hour) {
+          m_last_tm.tm_hour = m_tm.tm_hour;
+          hour->update();
+        }
+        if (m_tm.tm_min != m_last_tm.tm_min) {
+          m_last_tm.tm_min = m_tm.tm_min;
+          minutes->update();
+        }
+        if (m_tm.tm_sec != m_last_tm.tm_sec) {
+          m_last_tm.tm_sec = m_tm.tm_sec;
+          seconds->update();
+        }
       }
     }
 
   protected:
+    struct gtm m_tm;
+    struct gtm m_last_tm;
     tmr10ms_t lastRefresh = 0;
-    int8_t lastSecond;
     NumberEdit* year = nullptr;
     NumberEdit* month = nullptr;
     NumberEdit* day = nullptr;
     NumberEdit* hour = nullptr;
     NumberEdit* minutes = nullptr;
     NumberEdit* seconds = nullptr;
 
+    int8_t daysInMonth()
+    {
+      static const int8_t dmon[] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
+      int16_t year = TM_YEAR_BASE + m_tm.tm_year;
+      int8_t days_in_month = dmon[m_tm.tm_mon];
+      if ((m_tm.tm_mon == 1) && (((year % 4 == 0) && (year % 100 != 0)) || (year%400==0)))
+        days_in_month += 1;
+      return days_in_month;
+    }
+
+    void setDaysInMonth()
+    {
+        if (day) {
+            day->setMax(daysInMonth());
+            if (m_tm.tm_mday > day->getMax()) {
+                // Update stored day value if > actual days in month
+                // Will be written to RTC via SET_LOAD_DATETIME call after returning
+                // UI will update on next iteration of checkEvents
+                m_tm.tm_mday = day->getMax();
+            }
+        }
+    }
+
     void build()
     {
       setFlexLayout();
       FlexGridLayout grid(col_four_dsc, row_dsc, 2);
 
-      auto line = newLine(&grid);
+      gettime(&m_tm);
+      memcpy(&m_last_tm, &m_tm, sizeof(m_tm));
```

Comment:
you should be able to use a normal assignment here

Reply:
Thanks, my C++ is a bit rusty. I've made this change.

Reply:
That's what reviews are there for.